pkg/chains/storage/oci/attestation.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 The Tekton Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oci
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/in-toto/in-toto-golang/in_toto"
+	"github.com/pkg/errors"
+	"github.com/sigstore/cosign/v2/pkg/oci/mutate"
+	ociremote "github.com/sigstore/cosign/v2/pkg/oci/remote"
+	"github.com/sigstore/cosign/v2/pkg/oci/static"
+	"github.com/sigstore/cosign/v2/pkg/types"
+	"github.com/tektoncd/chains/pkg/chains/storage/api"
+	"github.com/tektoncd/chains/pkg/config"
+	"knative.dev/pkg/logging"
+)
+
+var (
+	_ api.Storer[name.Digest, in_toto.Statement] = &AttestationStorer{}
+)
+
+// AttestationStorer stores in-toto Attestation payloads in OCI registries.
+type AttestationStorer struct {
+	// repo configures the repo where data should be stored.
+	// If empty, the repo is inferred from the Artifact.
+	repo *name.Repository
+	// remoteOpts are additional remote options (i.e. auth) to use for client operations.
+	remoteOpts []remote.Option
+}
+
+func NewAttestationStorer(cfg config.Config) (*AttestationStorer, error) {
+	var repo *name.Repository
+	if r := cfg.Storage.OCI.Repository; r != "" {
+		resolved, err := name.NewRepository(r)
+		if err != nil {
+			return nil, errors.Wrapf(err, "%s is not a valid repository", r)
+		}
+		repo = &resolved
+	}
+
+	return &AttestationStorer{
+		repo: repo,
+	}, nil
+}
+
+func (s *AttestationStorer) Store(ctx context.Context, req *api.StoreRequest[name.Digest, in_toto.Statement]) (*api.StoreResponse, error) {
+	logger := logging.FromContext(ctx)
+	// upload an attestation for each subject
+	logger.Info("Starting to upload attestations to OCI ...")
+	for _, subj := range req.Payload.Subject {
```

Comment:
Is looping over the subjects right? The for-loop only references `subj` for logging purposes.

Reply:
Good catch! This was copy-pasted from the outer loop. Removed.